pep-9999.rst

```diff
@@ -0,0 +1,1037 @@
+files only within the directories of the ``sysconfig``
+scheme in which they would create files**. This permits a
+distributor of a Python interpreter to set up two
+directories, one for its own managed packages, and one for
+unmanaged packages installed by the end user, and ensure
+that installing unmanaged packages will not delete (or
+overwrite) files owned by the external package manager.
+
+Rationale
+=========
+
+As described in detail in the next section, the first
+behavior change involves creating a marker file named
+``EXTERNALLY-MANAGED``, whose presence indicates that
+non-virtual-environment package installations are managed by
+some means external to Python, such as a distribution's
+package manager. This file is specified
+to live in the ``stdlib`` directory in the default
+``sysconfig`` scheme, which marks the interpreter /
+installation as a whole, not a particular location on
+``sys.path``. The reason for this is that, as identified
+above, there are two related problems that risk breaking an
+externally-managed Python: you can install an incompatible
+new version of a package system-wide (e.g., with ``sudo pip
+install``), and you can install one locally, but in a
+location that is loaded by the standard Python command
+(e.g., with ``pip install --user``). If the marker file were
+in the system-wide ``site-packages`` directory, it would not
+clearly apply to the second case. The `Rejected Ideas`_
+section has further discussion of possible locations.
+
+The second behavior change takes advantage of the existing
+``sysconfig`` setup in distributions that have already
+encountered this class of problem, and specifically
+addresses the problem of a Python-specific package manager
+deleting or overwriting files that are owned by an external
+package manager.
+
+The changed behavior in this PEP is intended to "do the
+right thing" for as many use cases as possible. For the
+following representative use cases, we analyze the two
+behaviors that could be changed by this PEP: can you run a
+Python-specific installer tool like ``pip install`` by
+default, and if you do run such a tool, should it be willing
+to delete packages shipped by the external package manager?
+For simplicity, this section refers to pip's behavior,
+although it applies equivalently to any other
+Python-specific package management tool.
+
+1. A standard unpatched CPython, without any special
+   configuration of or patches to ``sysconfig`` and without
+   a marker file. The behavior does not change:
+
+   * Can you ``pip install`` without an override? Currently
+     yes; after this PEP yes.
+
+   * Can it delete externally-installed packages? Currently
+     yes; after this PEP yes. Because the ``sysconfig``
+     scheme used by any externally-installed packages is the
+     same scheme used by ``pip install`` (which is to say,
+     probably any externally-installed packages were
+     installed by pip too), there is no distinction.
```

Comment:
Is this desirable? I can't say I'd want to compile CPython for my Debian machine and then have break my system Python. Why wouldn't it respect the distribution hints?

Reply:
If you install upstream CPython and `sudo /usr/local/bin/python -m pip uninstall` something, it's not going to see distribution-installed things in `/usr/lib/python3/dist-packages` at all, i.e., "externally-installed packages" in this context doesn't mean distribution-installed packages because there is no relevant distribution. For the same reason it doesn't respect the distribution EXTERNALLY-MANAGED file because it doesn't see that file at all.

(If you install upstream CPython with, idk, `./configure -prefix=/usr` then it will see the EXTERNALLY-MANAGED file and the answer to at least the first question is yes->no, but also if you install upstream CPython into /usr, you're doing something extremely strange.)

Should I rephrase this to say something like "irrelevant because there is no such thing as externally-installed packages in this ctext"?

Reply:
Rephrased this a bit, see if it's clearer.